command/agent/config.go

```diff
@@ -668,22 +668,77 @@ func (r *Resources) CanParseReserved() error {
 	return err
 }
 
+// devModeConfig holds the config for the -dev flag
+type devModeConfig struct {
+	defaultMode bool
+	connectMode bool
+}
+
+// newDevModeConfig parses the optional string value of the -dev flag
+func newDevModeConfig(s string) (*devModeConfig, error) {
+	if s == "" {
+		return nil, nil // no -dev flag
+	}
+	mode := &devModeConfig{}
+	modeFlags := strings.Split(s, ",")
+	for _, modeFlag := range modeFlags {
+		switch modeFlag {
+		case "true": // -dev flag with no params
+			mode.defaultMode = true
+		case "connect":
+			if runtime.GOOS != "linux" {
+				// strictly speaking -dev=connect only binds to the
+				// non-localhost interface, but given its purpose
+				// is to support a feature with network namespaces
+				// we'll return an error here rather than let the agent
+				// come up and fail unexpectedly to run jobs
+				return nil, fmt.Errorf("-dev=connect is only supported on linux")
+			}
+			mode.connectMode = true
+		default:
+			return nil, fmt.Errorf("invalid -dev flag")
+		}
+	}
+	return mode, nil
+}
+
+func (mode *devModeConfig) networkConfig() (string, string) {
+	if runtime.GOOS == "darwin" {
+		return "127.0.0.1", "lo0"
+	}
+	if mode != nil && mode.connectMode {
+		// if we hit either of the errors here we're in a weird situation
+		// where syscalls to get the list of network interfaces are failing.
+		// rather than throwing errors, we'll fall back to the default.
+		iface, err := template.Parse(
+			`{{ GetDefaultInterfaces | limit 1 | attr "name" }}`)
```

Comment:
I have not seen them in practice... I'm not sure what it would take to get into that situation, but after switching off the sockaddr template as you've suggested I'm realizing we have the same UX problem if there are no public interfaces.

So yeah, let's fix that. There's no logger configured yet by the time we get here, but we can send an error back up to the caller of `newDevModeConfig` and report it at the command line.

Reply:
Made that fix in 67f0b2c